File: examples/rust/src/bin/02_mars_metadata.rs
//! Example 02 — MARS-namespaced metadata
//!
//! Shows how to attach ECMWF MARS vocabulary keys to a message and to
//! individual objects, then read them back after decoding.
//!
//! - Per-object MARS keys live in `base[i]["mars"]`.
//! - Each base entry holds ALL metadata for that object independently.
//!
//! The library is vocabulary-agnostic: it stores and returns whatever keys
//! you put in. Meaning is assigned by the application layer.

use std::collections::BTreeMap;

use ciborium::Value;
use tensogram_core::{
    decode, decode_metadata, encode, ByteOrder, DataObjectDescriptor, DecodeOptions, Dtype,
    EncodeOptions, GlobalMetadata,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // ── Per-object MARS keys → base[0]["mars"] ────────────────────────────────
    //
    // Each base entry holds ALL metadata for one object.  Keys shared across
    // objects are simply repeated in each entry — the library does not track
    // what is common vs varying.
    let mars_obj0 = Value::Map(vec![
        (Value::Text("class".into()), Value::Text("od".into())),
        (Value::Text("date".into()), Value::Text("20260401".into())),
        (Value::Text("step".into()), Value::Integer(6.into())),
        (Value::Text("time".into()), Value::Text("0000".into())),
        (Value::Text("type".into()), Value::Text("fc".into())),
        (Value::Text("levtype".into()), Value::Text("sfc".into())),
        (Value::Text("param".into()), Value::Text("2t".into())),
    ]);

    let mut obj0_base = BTreeMap::new();
    obj0_base.insert("mars".to_string(), mars_obj0);

    // ── Build descriptor and global metadata ──────────────────────────────────
    let desc = DataObjectDescriptor {
        obj_type: "ntensor".to_string(),
        ndim: 2,
        shape: vec![721, 1440], // 0.25-degree global grid
        strides: vec![1440, 1],
        dtype: Dtype::Float32,
        byte_order: ByteOrder::Big,
        encoding: "none".to_string(),
        filter: "none".to_string(),
        compression: "none".to_string(),
        params: BTreeMap::new(), // encoding params only — no MARS keys here
        hash: None,
    };

    let global_meta = GlobalMetadata {
        version: 2,
        base: vec![obj0_base],
        ..Default::default()
    };

    let data = vec![0u8; 721 * 1440 * 4]; // zeros stand in for real values
    let message = encode(&global_meta, &[(&desc, &data)], &EncodeOptions::default())?;

    // ── Read back metadata only (no payload decode) ───────────────────────────
    //
    // decode_metadata() parses the binary header and CBOR section, then stops.
    // No object bytes are read or allocated. Useful for filtering/listing.
    let meta = decode_metadata(&message)?;

    // ── Navigate the MARS namespace ───────────────────────────────────────────
    fn get_mars_key<'a>(map: &'a Value, key: &str) -> Option<&'a str> {
        if let Value::Map(entries) = map {
            for (k, v) in entries {
                if matches!(k, Value::Text(s) if s == key) {
                    if let Value::Text(t) = v {
                        return Some(t);
                    }
                }
            }
        }
        None
    }

    let mars = &meta.base[0]["mars"];
    println!("Object 0 (base mars):");
    println!("  class   = {}", get_mars_key(mars, "class").unwrap_or("?"));
    println!("  date    = {}", get_mars_key(mars, "date").unwrap_or("?"));
    println!("  type    = {}", get_mars_key(mars, "type").unwrap_or("?"));
    println!("  step    = {:?}", {
        if let Value::Map(e) = mars {
            e.iter()
                .find(|(k, _)| matches!(k, Value::Text(s) if s == "step"))
                .map(|(_, v)| v)
        } else {
            None
        }
    });

    // ── Full decode ───────────────────────────────────────────────────────────
    let (meta2, objects) = decode(&message, &DecodeOptions::default())?;

    // Per-object MARS keys are in base[i]["mars"]
    let obj_mars = &meta2.base[0]["mars"];
    println!("Object 0 (decoded mars):");
    println!(
        "  param   = {}",
        get_mars_key(obj_mars, "param").unwrap_or("?")
    );
    println!(
        "  levtype = {}",
        get_mars_key(obj_mars, "levtype").unwrap_or("?")
    );
    println!("  shape   = {:?}", objects[0].0.shape);

    assert_eq!(objects[0].1, data);
    println!("\nFull decode OK.");

    Ok(())
}

File: examples/typescript/02_mars_metadata.ts
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

/**
 * Example 02 — Per-object metadata with the MARS vocabulary (TypeScript)
 *
 * Mirrors `examples/python/02_mars_metadata.py`.
 *
 * Each data object gets its own `base[i]` entry with application metadata.
 * This example uses the ECMWF MARS vocabulary for concreteness; the same
 * pattern works with any namespace (CF, BIDS, DICOM, custom). See
 * `02b_generic_metadata.ts` for a non-MARS example.
 *
 * `getMetaKey` looks up dotted paths using first-match semantics;
 * `computeCommon` extracts keys that are shared across every object's
 * metadata.
 */

import {
  computeCommon,
  decode,
  encode,
  getMetaKey,
  init,
  type DataObjectDescriptor,
  type GlobalMetadata,
} from '@ecmwf/tensogram';

async function main(): Promise<void> {
  await init();

  const shape = [2, 3] as const;
  const strides = [3, 1] as const;
  const descriptor: DataObjectDescriptor = {
    type: 'ntensor',
    ndim: 2,
    shape: [...shape],
    strides: [...strides],
    dtype: 'float32',
    byte_order: 'little',
    encoding: 'none',
    filter: 'none',
    compression: 'none',
  };

  const temperature = new Float32Array([273.15, 274.0, 275.0, 276.0, 277.0, 278.0]);
  const wind = new Float32Array([0.0, 1.0, 2.0, 3.0, 4.0, 5.0]);

  const metadata: GlobalMetadata = {
    version: 2,
    base: [
      {
        mars: {
          class: 'od', type: 'fc', stream: 'oper',
          date: '20260401', time: '1200',
          param: '2t', levtype: 'sfc',
        },
      },
      {
        mars: {
          class: 'od', type: 'fc', stream: 'oper',
          date: '20260401', time: '1200',
          param: '10u', levtype: 'sfc',
        },
      },
    ],
    _extra_: { source: 'ifs-cycle49r2' },
  };

  const msg = encode(metadata, [
    { descriptor, data: temperature },
    { descriptor, data: wind },
  ]);
  console.log(`Message size: ${msg.byteLength} bytes`);

  const result = decode(msg);
  try {
    const meta = result.metadata;

    // Dotted-path lookup across base[i] with first-match semantics.
    console.log(`\nmars.param (first match): ${getMetaKey(meta, 'mars.param') as string}`);
    console.log(`mars.class (first match): ${getMetaKey(meta, 'mars.class') as string}`);
    console.log(`_extra_.source:           ${getMetaKey(meta, '_extra_.source') as string}`);
    console.log(`missing key:              ${String(getMetaKey(meta, 'does.not.exist'))}`);

    // computeCommon() extracts top-level keys whose value is identical across
    // every base entry. The `mars` sub-objects here differ by `param`, so at
    // the top level they are NOT equal — nothing comes out as common. This is
    // by design: commonality is a whole-value check, not a recursive merge.
    const common = computeCommon(meta);
    console.log(`\ncomputeCommon() (no shared top-level key expected here):`);
    console.log(`  keys: [${Object.keys(common).join(', ')}]`);

    // To see computeCommon() actually return a shared key, flatten the MARS
    // sub-object into top-level keys at the base level.
    const flatMeta = {
      version: 2,
      base: [
        { class: 'od', type: 'fc', param: '2t' },
        { class: 'od', type: 'fc', param: '10u' },
      ],
    };
    console.log(`  on a flat metadata: keys = [${Object.keys(computeCommon(flatMeta)).join(', ')}]`);

    console.log(`\nPer-object param values:`);
    for (let i = 0; i < result.objects.length; i++) {
      const entry = meta.base?.[i];
      const param = (entry?.mars as Record<string, unknown> | undefined)?.param;
      console.log(`  object[${i}] mars.param = ${String(param)}`);
    }
  } finally {
    result.close();
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
